modules/javafx.graphics/src/main/java/com/sun/javafx/css/CalculatedValue.java

```diff
@@ -88,6 +88,18 @@ public boolean isRelative() {
         return true;
     }
 
+    @Override
+    public int hashCode() {
+        int h = relative ? 1 : 0;
+        if (origin != null) {
+            h = 31 * h + origin.hashCode();
+        }
+        if (value != null) {
+            h = 31 * h + value.hashCode();
+        }
+        return h;
+    }
```

Comment:
Just an example, but wouldn't: `Objects.hash(relative, origin, value)` here work just as well?

Reply:
Honestly, I would just turn this class into a `record` and everything will be taken care of. I'm not sure it is within the scope of the fix.

Reply:
Such a change is definitely out of scope. More importantly, we can't use `record` anywhere in JavaFX until we bump the minimum version of the JDK needed to run JavaFX. I plan to start a discussion on the openjfx-dev mailing list, since it has been something I've wanted to do for a while now.

Reply:
@hjohn: yes, but at a price: Object.hash(Object ...) incurs overhead by creating a temporary Object[] + boxing of a Boolean.

Reply:
Ah, I didn't realize you checked how this is optimized by the JIT.

Reply:
In order to reduce collisions, the hash of each component is typically added to `h * 31` even when that hash is 0, whereas you skip the `h = 31 * h` in the case of null. It might not be a problem in practice, since value and origin are unlikely to collide, being of different types, but you might want to look at it.

In any case, I need to time to look at it, which I won't have until after JavaFX 19 RDP1, so let's leave this until then.

Reply:
you bring a good point, Kevin, thank you!

Reply:
I would use `Boolean.hashCode(relative);` for a `boolean`.

Kevin, I checked Effective Java 3rd Edition and it also says to use 0 (or some other constant) for `null`.

Reply:
Using 0 for a null component of the hash is fine, as long as the accum variable is initialized to something > 0, and as long as you accumulate all values, even a 0. By which I mean that the following is not OK:

```
    int h = 1;
    if (comp1 != null) h = 31 * h + comp1.hashCode();
    if (comp2 != null) h = 31 * h + comp2.hashCode();
```

nor is this:

```
    int h = (comp1 == null) ? 0 : comp1.hashCode();
    h = 31 * h + (comp2 == null) ? 0 : comp2.hashCode();
```


but this is:

```
    int h = 1;
    h = 31 * h + (comp1 == null) ? 0 : comp1.hashCode();
    h = 31 * h + (comp2 == null) ? 0 : comp2.hashCode();
```

Anyway, the latest change in this PR is good.